src/details/ArborX_DetailsUnionFind.hpp

```diff
@@ -171,37 +146,28 @@ struct UnionFind
     // until there is no data race on the parent.
     // ```
 
-    // ##### ECL license (see LICENSE.ECL) #####
     int vstat = representative(i);
     int ostat = representative(j);
 
-    bool repeat;
-    do
+    if constexpr (Serial)
     {
-      repeat = false;
-      if (vstat != ostat)
+      if (vstat < ostat)
+        _labels(ostat) = vstat;
+      else
+        _labels(vstat) = ostat;
+    }
+    else
+    {
+      while (vstat != ostat)
       {
-        int ret;
         if (vstat < ostat)
-        {
-          if ((ret = Kokkos::atomic_compare_exchange(&_labels(ostat), ostat,
-                                                     vstat)) != ostat)
-          {
-            ostat = ret;
-            repeat = true;
-          }
-        }
+          ostat =
+              Kokkos::atomic_compare_exchange(&_labels(ostat), ostat, vstat);
```

Comment:
This makes things more readable.

Reply:
There is one minor details that I missed here which would do one extra unnecessary while iteration. Let me try to fix it.

Reply:
I see no difference between these two:
```diff
-  ostat = Kokkos::atomic_compare_exchange(&_labels(ostat), ostat, vstat);
+ ostat = _labels(Kokkos::atomic_compare_exchange(&_labels(ostat), ostat, vstat));
```
In the first case, when replacement is done, we would not hit `ostat == vstat` immediately, rather we'll do another loop iteration with another atomic exchange. In the second, we won't. However, I see no difference in CUDA performance among all three versions (original, new, and new with labels). So I'm fine leaving it as is.